chain/messagepool/messagepool.go

```diff
@@ -863,7 +847,20 @@ func (mp *MessagePool) checkBalance(ctx context.Context, m *types.SignedMessage,
 	return nil
 }
 
-func (mp *MessagePool) addTs(ctx context.Context, m *types.SignedMessage, curTs *types.TipSet, local, untrusted bool) (bool, error) {
+func (mp *MessagePool) addTs(ctx context.Context, m *types.SignedMessage, local, untrusted bool) (bool, error) {
+	//ensures that we have a consistent view of the state
+	mp.transactionLk.Lock()
```

Comment:
It's not immediately clear to me why we need the `transactionLk` here, and in some of the subsequent methods, since we aren't modifying `curTs`. I can totally believe it's necessary, but can you please talk me through it?

Reply:
Hey! That's a good question! When the tip set changes, the message pool is pruned to remove any messages that were in the blocks:

```
// caller must hold transactionLk and stateLk
func (mp *MessagePool) headChange(ctx context.Context, revert []*types.TipSet, apply []*types.TipSet) error {

  repubTrigger := false
  rmsgs := make(map[address.Address]map[uint64]*types.SignedMessage)
  add := func(m *types.SignedMessage) {
    s, ok := rmsgs[m.Message.From]
    if !ok {
      s = make(map[uint64]*types.SignedMessage)
      rmsgs[m.Message.From] = s
    }
    s[m.Message.Nonce] = m
  }
  rm := func(from address.Address, nonce uint64) {
    s, ok := rmsgs[from]
    if !ok {
      mp.remove(ctx, from, nonce, true)
      return
    }

    if _, ok := s[nonce]; ok {
      delete(s, nonce)
      return
    }

    mp.remove(ctx, from, nonce, true)
  }
...
  for _, ts := range apply {
    mp.curTs = ts

    for _, b := range ts.Blocks() {
      bmsgs, smsgs, err := mp.api.MessagesForBlock(ctx, b)
      if err != nil {
        xerr := xerrors.Errorf("failed to get messages for apply block %s(height %d) (msgroot = %s): %w", b.Cid(), b.Height, b.Messages, err)
        log.Errorf("error retrieving messages for block: %s", xerr)
        merr = multierror.Append(merr, xerr)
        continue
      }

      for _, msg := range smsgs {
        rm(msg.Message.From, msg.Message.Nonce)
        maybeRepub(msg.Cid())
      }

      for _, msg := range bmsgs {
        rm(msg.From, msg.Nonce)
        maybeRepub(msg.Cid())
      }
    }
  }
```
we want to wrap changes to tipset and/or messages in a pair of transaction and state locks because changing the message pool requires a coordination lock on the tipset because the tipset changes the message pool